lib/runner.js

```diff
@@ -44,67 +46,84 @@ Runner.prototype.addAfterHook = function (title, cb) {
 	this.tests.after.push(new Test(title, cb));
 };
 
-Runner.prototype.concurrent = function (tests, cb) {
-	each(tests, function (test, i, next) {
-		test.run(function (err, duration) {
-			if (err) {
-				this.stats.failCount++;
-			}
-
-			this.results.push({
-				duration: duration,
-				title: test.title,
-				error: err
-			});
-
-			this.emit('test', err, test.title, duration);
-			next();
-		}.bind(this));
-	}.bind(this), cb);
+Runner.prototype.concurrent = function (tests) {
+	return each(tests, function (test, i, next) {
+		test.run()
+			.catch(function () {
+				return;
+			})
+			.then(function () {
+				if (test.assertError) {
+					this.stats.failCount++;
+				}
+
+				this.results.push({
+					duration: test.duration,
+					title: test.title,
+					error: test.assertError
+				});
+
+				this.emit('test', test.assertError, test.title, test.duration);
+				next();
+			}.bind(this));
+	}.bind(this));
 };
 
-Runner.prototype.serial = function (tests, cb) {
-	eachSerial(tests, function (test, next) {
-		test.run(function (err, duration) {
-			if (err) {
-				this.stats.failCount++;
-			}
-
-			this.results.push({
-				duration: duration,
-				title: test.title,
-				error: err
-			});
-
-			this.emit('test', err, test.title, duration);
-			next();
-		}.bind(this));
-	}.bind(this), cb);
+Runner.prototype.serial = function (tests) {
+	return eachSerial(tests, function (test, next) {
+		test.run()
+			.catch(function () {
+				return;
+			})
+			.then(function () {
+				if (test.assertError) {
+					this.stats.failCount++;
+				}
+
+				this.results.push({
+					duration: test.duration,
+					title: test.title,
+					error: test.assertError
+				});
+
+				this.emit('test', test.assertError, test.title, test.duration);
+				next();
+			}.bind(this));
+	}.bind(this));
 };
 
-Runner.prototype.run = function (cb) {
+Runner.prototype.run = function () {
 	var concurrent = this.tests.concurrent;
 	var serial = this.tests.serial;
 	var before = this.tests.before;
 	var after = this.tests.after;
 
-	// TODO: refactor this bullshit
-	this.serial(before, function () {
-		if (this.stats.failCount > 0) {
-			return this.end(cb);
-		}
-
-		this.serial(serial, function () {
-			this.concurrent(concurrent, function () {
-				this.serial(after, function () {
-					this.end(cb);
-				}.bind(this));
-			}.bind(this));
-		}.bind(this));
-	}.bind(this));
-};
+	var self = this;
 
-Runner.prototype.end = function (cb) {
-	this.stats.passCount = this.stats.testCount - this.stats.failCount;
-	cb(this.stats, this.results);
+	return this.serial(before)
+		.then(function () {
+			if (self.stats.failCount > 0) {
+				return Promise.reject();
+			}
+		})
+		.then(function () {
+			return self.serial(serial);
+		})
+		.then(function () {
+			return self.concurrent(concurrent);
+		})
+		.then(function () {
+			return self.serial(after);
+		})
+		.catch(function () {
+			return;
```

Comment:
Oh, right, I am just checking notes from the PR, not from the files themselves.

Because previously, when error happened, `this.end` executed and that's it. So that's exactly what I did here. In case of an error, just exit.